test/sequence/identity_view.cpp

```diff
@@ -0,0 +1,184 @@
+        BOOST_TEST(at_key<abstract>(t) == 2);
+
+        BOOST_STATIC_ASSERT((
+            boost::is_same<boost::fusion::result_of::value_at_key<transformed_type, int>::type, char>::value));
+        BOOST_STATIC_ASSERT((
+            boost::is_same<boost::fusion::result_of::value_at_key<transformed_type, double>::type, std::string>::value));
+        BOOST_STATIC_ASSERT((
+            boost::is_same<boost::fusion::result_of::value_at_key<transformed_type, abstract>::type, int>::value));
+
+        std::cout << t << std::endl;
+
+        BOOST_STATIC_ASSERT((boost::fusion::result_of::has_key<transformed_type, int>::value));
+        BOOST_STATIC_ASSERT((boost::fusion::result_of::has_key<transformed_type, double>::value));
+        BOOST_STATIC_ASSERT((boost::fusion::result_of::has_key<transformed_type, abstract>::value));
+        BOOST_STATIC_ASSERT((!boost::fusion::result_of::has_key<transformed_type, std::string>::value));
+
+        std::cout << deref_data(begin(t)) << std::endl;
+        std::cout << deref_data(boost::fusion::next(begin(t))) << std::endl;
+
+        BOOST_TEST(deref_data(begin(t)) == 'X');
+        BOOST_TEST(deref_data(boost::fusion::next(begin(t))) == "Men");
+        BOOST_TEST(deref_data(boost::fusion::next(next(begin(t)))) == 2);
+
+        BOOST_STATIC_ASSERT((boost::is_same<boost::fusion::result_of::key_of<boost::fusion::result_of::begin<transformed_type>::type>::type, int>::value));
+        BOOST_STATIC_ASSERT((boost::is_same<boost::fusion::result_of::key_of<boost::fusion::result_of::next<boost::fusion::result_of::begin<transformed_type>::type>::type>::type, double>::value));
+        BOOST_STATIC_ASSERT((boost::is_same<boost::fusion::result_of::key_of<boost::fusion::result_of::next<boost::fusion::result_of::next<boost::fusion::result_of::begin<transformed_type>::type>::type>::type>::type, abstract>::value));
+        BOOST_STATIC_ASSERT((boost::is_same<boost::fusion::result_of::value_of_data<boost::fusion::result_of::begin<transformed_type>::type>::type, char>::value));
+        BOOST_STATIC_ASSERT((boost::is_same<boost::fusion::result_of::value_of_data<boost::fusion::result_of::next<boost::fusion::result_of::begin<transformed_type>::type>::type>::type, std::string>::value));
+        BOOST_STATIC_ASSERT((boost::is_same<boost::fusion::result_of::value_of_data<boost::fusion::result_of::next<boost::fusion::result_of::next<boost::fusion::result_of::begin<transformed_type>::type>::type>::type>::type, int>::value));
+
+        // Test random access interface.
+        pair<int, char> a = at_c<0>(t); (void) a;
+        pair<double, std::string> b = at_c<1>(t);
+        pair<abstract, int> c = at_c<2>(t);
+        (void)c;
+
+        typedef boost::fusion::result_of::begin<transformed_type>::type first;
+        typedef boost::fusion::result_of::next<first>::type second;
+        typedef boost::fusion::result_of::next<second>::type third;
+
+        // TODO: why is a boost::fusion::pair<int, char>&& ??
```

Comment:
The problem is that `identity` will never return a value, it will only return a reference. Moreover - in C++03 it will not compile.

Reply:
I see 3 variants to resolve it:
1. To add a special flag on the `transform_view` side that will allow to ignore `F` when `value_of` or `value_at`.
2. To patch all Sequence's and Iterator's methods (`value_of`, `value_at`, `begin`, `end`, etc..). Add ability to override implementations selected by `tag_of`.
3. To implement `identity_view` manually.. Don't use `transform_view` and some identity funobject.

Reply:
4. To implement specific identity funobject(without rvalue) manually. Don't use `std::identity` or `boost::identity`, even after C++11 standards, we don't need it cause of fusion don't use rvalues.

Reply:
> To add a special flag on the `transform_view` side that will allow to ignore `F` when `value_of` or `value_at`.

I suppose that ability to set "specific funobject for **prvalue**" will be better. And then, templated iface of `transform_view` will be like this:
```
template <typename Sequence, typename F1, typename FPRV = F1>
struct transform_view;
```

And then we can use it on `identity_view` side like this:
```
struct prval_identity {
    template<typename T>
    T operator() (T val) const;
};

transform_view<Sequence, boost::identity, prval_identity> tv(seq, boost::identity(), prval_identity());
```


> To patch all Sequence's and Iterator's methods (`value_of`, `value_at`, `begin`, `end`, etc..). Add ability to override implementations selected by `tag_of`.

Too many changes patch.

> To implement `identity_view` manually.. Don't use `transform_view` and some identity funobject.

Guaranteed to work, but we will leave this variant extreme.

> To implement specific identity funobject(without rvalue) manually. Don't use `std::identity` or `boost::identity`, even after C++11 standards, we don't need it cause of fusion don't use rvalues.

It's wrong. We need identity with rvalues support, we can't spoil, for example, this sequence: `vector<std::string, std::string&&>`.
We really need identity funobject, that won't spoil **prvalues**, but its impossible, for example:
```
// It will not spoil prvalues, but it will be compile time error for other value categories:
// - error: call of '(identity) (int&)' is ambiguous
// - error: call of '(identity) (const int&)' is ambiguous
struct identity 
{
   template<typename T>
   T operator() (T val) const;

   template<typename T>
   T& operator() ( T& val) const;
};
```

Reply:
> 4. To implement specific identity funobject(without rvalue) manually. Don't use `std::identity` or `boost::identity`, even after C++11 standards, we don't need it cause of fusion don't use rvalues.

More correctly, C++03 does not have r-values...

Makes me think that we need to decide when to move on and leave C++03. I think the time is now. I think we can do this incrementally. I also think that new features should be free from having to support C++03. It's best to just forget about C++03 for this extension. Test for C++ version in the code and disable entire headers and tests when 03 is detected. Of course, this should be documented as well.

Reply:
@djowel, OK.
What do you think about this strategy?

> I suppose that ability to set "specific funobject for **prvalue**" will be better. And then, templated iface of `transform_view` will be like this:
> ```
> template <typename Sequence, typename F1, typename FPRV = F1>
> struct transform_view;
> ```
> 
> And then we can use it on `identity_view` side like this:
> ```
> struct prval_identity {
>     template<typename T>
>     T operator() (T val) const;
> };
> 
> transform_view<Sequence, boost::identity, prval_identity> tv(seq, boost::identity(), prval_identity());
> ```

If it unacceptable, i will recreate `identity_view` manually.

Reply:
It's looking like the limitations of the implementation is seeping out the API.

Reply:
> 4. To implement specific identity funobject(without rvalue) manually. Don't use `std::identity` or `boost::identity`, even after C++11 standards, we don't need it cause of fusion don't use rvalues.

5. Implement identity funobject in a detail namespace, and then add `boost::result_of` specialization for it.
Just like... https://godbolt.org/z/6Yne5cTb4
I know it looks like a dirty crutch, but it works great for us.

Reply:
> 5\. Implement identity funobject in a detail namespace, and then add `boost::result_of` specialization for it.
>     Just like... https://godbolt.org/z/6Yne5cTb4
>     I know it looks like a dirty crutch, but it works great for us.

Go for it

Reply:
@djowel docs was not checked. I have no ability to compile them now.

Reply:
Look at that! Clean build!